app/Http/Controllers/Administration/UpdateController.php

```diff
@@ -117,23 +122,33 @@ public function view(): View
 	 * The whole code around installation/upgrade/migration should
 	 * thoroughly be revised an refactored.
 	 */
-	public function migrate(Request $request): View
+	public function migrate(Request $request): View|Response
 	{
-		if (
-			AccessControl::is_admin() || AccessControl::noLogin() ||
-			AccessControl::log_as_admin($request['username'] ?? '', $request['password'] ?? '', $request->ip())
-		) {
+		// This conditional code makes use of lazy boolean evaluation: a || b does not execute b if a is true.
+		// 1. We check if we are logged in
+		// 2. if not, we check if the admin user is registered (if not login in such case)
+		// 3. We have an admin user set up, try to login with Legacy method: hash(username) + hash(password).
+		// 4. if Legacy login failed, try to login the normal way.
+		$isLoggedIn = Auth::check();
+		$isLoggedIn = $isLoggedIn || Authorization::loginAsAdminIfNotRegistered();
+		$isLoggedIn = $isLoggedIn || Legacy::loginAsAdmin($request->input('username', ''), $request->input('password', ''), $request->ip());
```

Comment:
I withdraw this comment. I thought it would be nice to implement a custom `LegacyUserProvider` which extends `EloquentUserProvider` and overwrites the `retrieveByCredentials` method. I falsely assumed that the problem could be fixed by adding an `orWhere` to the query builder to also include results for hashed user names. But obviously this cannot work for salted hashes :grinning:

Hence, we should keep it the way it is and hope that it will be gone soon.